cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -85,67 +87,115 @@ message Result {
+Does envoy also read from bpf maps for identity resolution ? if so we can resuse the same mechanism.
+
+### Flag to disable built-in proxy
+
+Adding a flag(`--disable-builtin-dns-proxy`) to cilium agent to disable the built in dns proxy allows standlone dns proxy to run its independent lifecycle. The current implmentation of cilium agent depends on the built-in dns proxy for restoring the endpoint for DNS rules. With the flag enabled we will need to decouple that dependency. Cilium agent can retrieve the DNS rules from in memory(as they have already been parsed from the file system).
+
+#### Pros
+
+* Allows cilium agent to be upgraded without breaking the DNS traffic.
+* Provides a freedom to allow any delegated dns proxy to be plugged in.
+* Reduces the resource utilization in cilium agent.
+
+#### Cons
+
+* Higher latency in DNS resolution as the DNS queries are now handled by a separate process.
+
 ### Q: Restoring L7 DNS policy when agent is unavailable
 
-In an event where SDP restarts when agent's gRPC service is unavailable, can / should SDP rely on information stored on the filesystem to restore per endpoint L7 policy ?
+In an event where SDP restarts when agent's gRPC service is unavailable. SDP can read from the filesystem and restore the policy. The state is stored in `ep_config.json` file. So the flow would be first trying to connect to agent and if it fails, read from the file.
 
+### Scenarios and Expected Behaviour
 
-## Future Milestones
+In an ideal scenario, both cilium agent and SDP should be interacting with each other through a biredirectional grpc stream. However, in an event where either of them restarts/upgrade/donwgrade, what should be the expected behavior ?
 
-* Plumb toFQDN policy for new names when agent is unavailable
+#### Cilium Agent is down, SDP is running
+
+* SDP should be able to proxy requests and enfore L7 DNS policy based on the existing bpf maps and in memory DNS rules. The datapath of already configured policies will work. Any new mappings will not be updated until cilium agent is back up.
+* In case SDP restarts while cilium agent is down, the SDP should be able to read the rules from the filesystem and restore them.
+
+#### Cilium Agent is running, SDP is down/restarting
+
+* In case Cilium Agent has the DNS proxy running as well, then DNS queries will be served.
+* Once SDP is starting up, it will connect to the agent and get the latest policy updates.
 
+#### SDP Upgrade(Given Cilium Agent is Running)
 
+* Cilium agent can keep serving the DNS queries through in built DNS proxy. In case, the inbult dns proxy is disabled, the SDP needs an upgrade path. This can be achieved using the `maxsurge` on the sdp. This can be future milestone.
```

Comment:
```suggestion
* Cilium agent can keep serving the DNS queries through in built DNS proxy. In case, the inbult dns proxy is disabled, the SDP needs an upgrade path. This can be achieved using the `maxsurge` upgrade.
```